mistral-search-engine.ipynb

```diff
@@ -0,0 +1,652 @@
+      ]
+     },
+     "execution_count": 16,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "function_result = tools_to_str(names_to_functions[function_name](**function_params))\n",
+    "function_result"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "\"Apple's Worldwide Developers Conference (WWDC) took place on June 10, 2024. The keynote focused mainly on artificial intelligence, with several announcements made. Unfortunately, I don't have real-time information, so I can't provide the specific details of the 13 biggest announcements mentioned in the article from The Verge.\\n\\nHowever, some general expectations before the event included the possibility of new apps for iPadOS such as a native calculator and new features for Apple Pencil 3. For macOS, it was anticipated that there would be AI features and smart tools integrated into native apps like Pages, Keynote, and Xcode.\\n\\nThe event also included video sessions and opportunities for developers to engage with Apple designers and engineers. There was an in-person experience at Apple Park for selected developers to watch the keynote, meet with Apple team members, and participate in special activities. Apple also continued its tradition of supporting the next generation of developers through the Swift Student Challenge.\""
+      ]
+     },
+     "execution_count": 17,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "messages.append(ChatMessage(role=\"tool\", name=function_name, content=function_result, tool_call_id=tool_call.id))\n",
+    "\n",
+    "response = client.chat(model=model, messages=messages)\n",
+    "response.choices[0].message.content\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Chat in a chain (cleaner user experience)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "messages = []\n",
+    "\n",
+    "while True:\n",
+    "    input_ = input(\"Ask: \")\n",
+    "    messages.append(ChatMessage(role=\"user\", content=input_))\n",
+    "    response = client.chat(model=model, messages=messages, tools=tools, tool_choice=\"force\")\n",
```

Comment:
Force is not an option, maybe 'any'?

And I think thats the only thing, everything else seems to work.

Reply:
yea sorry you're right I keep getting confused with force and any. Just made the change !